docs/dapp/sapphire/precompiles.md

```diff
@@ -0,0 +1,110 @@
+cheaper to perform: x25519 key derivation, and DeoxysII-based encryption
+and decryption.
+
+These can be called in much the same way as other precompiles by
+dispatching calls to specific well-known contract addresses, as
+described below. Input parameters should be packed into a contiguous
+memory region, with each piece of data padded as usual to 32 bytes.
+
+## X25519 Key Derivation
+
+### Overview
+
+* Precompile address: `0x0100000000000000000000000000000000000002`
+* Input format: `bytes32 public_key || bytes32 private_key`
+* Gas cost: 100,000
+
+### Example
+
+```solidity
+function derive_symmetric_key(bytes32 key_public, bytes32 key_private) private view returns (bytes32) {
+    bytes32[3] memory data;
+    data[0] = key_public;
+    data[1] = key_private;
+    assembly {
+        let success := staticcall(gas(), 0x0100000000000000000000000000000000000002, data, 0x40, add(data, 0x40), 0x20)
+        if iszero(success) {
+            revert(0, 0)
+        }
+    }
+    return data[2];
+}
+```
+
+## DeoxysII Encryption
+
+### Overview
+
+* Encryption precompile address: `0x0100000000000000000000000000000000000003`
+* Decryption precompile address: `0x0100000000000000000000000000000000000004`
+* Input format: `bytes32 key || bytes32 nonce || uint text_length || uint additional_data_length || text_or_ciphertext || additional_data`
+* Gas cost: 50,000 minimum plus 100 per byte of input
+
+:::caution
+
+Both the text and additional data should each be padded to a multiple of
+32 bytes.
+
+:::
+
+### Example
+
+```solidity
+function encrypt(bytes32 key, bytes32 nonce, bytes memory text, bytes memory additional_data) private view returns (bytes memory result) {
+    assembly {
+        let p := mload(0x40)
+        mstore(p, key)
+        mstore(add(p, 0x20), nonce)
+        let text_len := mload(text)
+        mstore(add(p, 0x40), text_len)
+        let ad_len := mload(additional_data)
+        mstore(add(p, 0x60), ad_len)
+        let i := 0
+        for {} lt(i, text_len) { i := add(i, 0x20) } {
+            mstore(add(add(p, 0x80), i), mload(add(add(text, 0x20), i)))
+        }
+        let ad_begin := add(add(p, 0x80), i)
+        for { i := 0 } lt(i, ad_len) { i := add(i, 0x20) } {
+            mstore(add(ad_begin, i), mload(add(add(additional_data, 0x20), i)))
+        }
+        mstore(0x40, add(ad_begin, i))
+        let out := mload(0x40)
+        let success := staticcall(gas(), 0x0100000000000000000000000000000000000003, p, sub(out, p), out, 1)
+        if iszero(success) {
+            revert(0, 0)
+        }
+        mstore(out, returndatasize())
+        returndatacopy(add(out, 0x20), 0, returndatasize())
+        mstore(0x40, add(add(out, 0x20), returndatasize()))
+        result := out
+    }
+}
+```
+
+## Library
+
+The examples above show how to call the precompiles directly. For a more
+comfortable experience, wrappers are also provided in the Sapphire
+repository, in `library/Sapphire.sol`. The wrappers have the same
```

Comment:
Where is this Sapphire repo? I cannot seem to find any solidity code in https://github.com/oasisprotocol/sapphire-paratime/. Please add a link to `library/Sapphire.sol` file.

Reply:
https://github.com/oasisprotocol/sapphire-paratime/pull/44